### PR TITLE
EventsControllerTest - tests for EventStreamSourceGraph stage [MARATHON-7317]

### PR DESCRIPTION
**Pull request by gosubpl**
This is work-in-progress. And my first PR to the `marathon` repo.
I am trying to understand what sort of tests could be useful for you. I have drafted two tests for the `EventStreamSorceGraph` stage. There should probably be some more, for the edge cases resulting from commands like `Standby` and `ElectedAsLeader`.
Please let me know if this is useful, and if yes - what additional tests you would you like me to add.
https://jira.mesosphere.com/browse/MARATHON-7317

### CODE REVIEW

File: src/test/scala/mesosphere/marathon/api/akkahttp/v2/EventsControllerTest.scala [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+        .fromGraph[MarathonEvent, NotUsed](
+          new EventStreamSourceGraph(fakeEventBus, 5, fakeClientIp)
+        )
+      val fut = sourceUnderTest.take(3).runWith(Sink.seq)
+      Thread.sleep(200) // avoid race with materialization of the stream
+      fakeEventBus.publish(fakeApiPostEvent)
+      Thread.sleep(200)
+      fakeEventBus.publish(fakeApiPostEvent)
+      val result = Await.result(fut, 5 seconds)
+      println(result)
```
Comment (ichernetsky) [inline src/test/scala/mesosphere/marathon/api/akkahttp/v2/EventsControllerTest.scala]: Could you please use `logger` here which can be inherited from `StrictLogging`?

File: src/test/scala/mesosphere/marathon/api/akkahttp/v2/EventsControllerTest.scala [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+      assert(result.tail.head == fakeApiPostEvent)
```
Comment (ichernetsky) [inline src/test/scala/mesosphere/marathon/api/akkahttp/v2/EventsControllerTest.scala]: Could you please use `shouldBe` here?

File: src/test/scala/mesosphere/marathon/api/akkahttp/v2/EventsControllerTest.scala [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+
+  "EventStreamSourceGraph" should {
+    "be a valid source" in {
+      val fakeEventBus: EventStream = new EventStream(system)
+      val sourceUnderTest = Source
+        .fromGraph[MarathonEvent, NotUsed](
+          new EventStreamSourceGraph(fakeEventBus, 5, fakeClientIp)
+        )
+      val fut = sourceUnderTest.take(3).runWith(Sink.seq)
+      Thread.sleep(200) // avoid race with materialization of the stream
```
Comment (timcharper) [inline src/test/scala/mesosphere/marathon/api/akkahttp/v2/EventsControllerTest.scala]: I have rewritten this graph to remove the race. Time-based tests are a recipe for indeterminism.

Check it out in this PR https://github.com/mesosphere/marathon/pull/5548

Comment (timcharper) [reply src/test/scala/mesosphere/marathon/api/akkahttp/v2/EventsControllerTest.scala]: (I'd just go ahead and remove this test :) The tests I wrote test the controller's usage of the stream, and covers this case)

File: src/test/scala/mesosphere/marathon/api/akkahttp/v2/EventsControllerTest.scala [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+      fakeEventBus.publish(fakeApiPostEvent)
+      Thread.sleep(200)
+      fakeEventBus.publish(fakeApiPostEvent)
+      val result = Await.result(fut, 5 seconds)
+      println(result)
+      assert(result.tail.head == fakeApiPostEvent)
+    }
+
+    "fail for queue overload" in {
+      val fakeEventBus: EventStream = new EventStream(system)
```
Comment (timcharper) [inline src/test/scala/mesosphere/marathon/api/akkahttp/v2/EventsControllerTest.scala]: Would you mind reworking this test on top of the changes here? https://github.com/mesosphere/marathon/pull/5548

This case is currently missing (what happens during event overflow) and it'd be great to get a test for it. :)

File: src/test/scala/mesosphere/marathon/api/akkahttp/v2/EventsControllerTest.scala [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+      val sourceUnderTest = Source
+        .fromGraph[MarathonEvent, NotUsed](
+          new EventStreamSourceGraph(fakeEventBus, 1, fakeClientIp)
+        )
+      val fut = sourceUnderTest.take(3).throttle(1, 1 second, 1, ThrottleMode.shaping).runWith(Sink.seq)
+
+      Thread.sleep(200) // avoid materialization race
+      var i: Int = 0
```
Comment (meln1k) [inline src/test/scala/mesosphere/marathon/api/akkahttp/v2/EventsControllerTest.scala]: How about `(1 to 20).foreach(fakeEventBus.publish(fakeApiPostEvent))`?

File: src/test/scala/mesosphere/marathon/api/akkahttp/v2/EventsControllerTest.scala (hunk 1 above)
Comment (gosubpl) [reply src/test/scala/mesosphere/marathon/api/akkahttp/v2/EventsControllerTest.scala]: Absolutely, will do!

File: src/test/scala/mesosphere/marathon/api/akkahttp/v2/EventsControllerTest.scala (hunk 2 above)
Comment (gosubpl) [reply src/test/scala/mesosphere/marathon/api/akkahttp/v2/EventsControllerTest.scala]: Yup

File: src/test/scala/mesosphere/marathon/api/akkahttp/v2/EventsControllerTest.scala (hunk 3 above)
Comment (gosubpl) [reply src/test/scala/mesosphere/marathon/api/akkahttp/v2/EventsControllerTest.scala]: Great news, thanks for the heads-up. Will remove the test.

File: src/test/scala/mesosphere/marathon/api/akkahttp/v2/EventsControllerTest.scala (hunk 4 above)
Comment (gosubpl) [reply src/test/scala/mesosphere/marathon/api/akkahttp/v2/EventsControllerTest.scala]: Yes, I will look into that this week-end.

File: src/test/scala/mesosphere/marathon/api/akkahttp/v2/EventsControllerTest.scala (hunk 5 above)
Comment (gosubpl) [reply src/test/scala/mesosphere/marathon/api/akkahttp/v2/EventsControllerTest.scala]: The whole test will be re-worked, see comments above.

**Comment by gosubpl** [comment]
Thanks for the comments everyone. Especially @timcharper , I will re-work the changes on top of your PR.

**Comment by gosubpl** [comment]
Thanks for the comments everyone. I am closing for now. If you ever need the overflow test in new setup, it is here: https://github.com/gosubpl/marathon/commit/67b2a16931dab81f79a69938358b3cc392e8f62c
